ppl/archive/index/writer.go

```diff
@@ -0,0 +1,152 @@
+package index
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/brimsec/zq/expr"
+	"github.com/brimsec/zq/field"
+	"github.com/brimsec/zq/microindex"
+	"github.com/brimsec/zq/pkg/iosrc"
+	"github.com/brimsec/zq/proc"
+	"github.com/brimsec/zq/proc/compiler"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"gopkg.in/tomb.v2"
+)
+
+func NewWriter(ctx context.Context, u iosrc.URI, def *Def) (*Writer, error) {
+	indexer, err := newIndexer(ctx, u, def)
+	if err != nil {
+		return nil, err
+	}
+	w := &Writer{
+		URI:     u,
+		Def:     def,
+		indexer: indexer,
+	}
+	w.WriteCloser = zbuf.BatchRecordWriter(w)
+	return w, nil
+}
+
+type Writer struct {
+	Def *Def
+	URI iosrc.URI
+	zbuf.WriteCloser
+
+	indexer *indexer
+}
+
+func (w *Writer) WriteBatch(batch zbuf.Batch) error {
+	select {
+	case w.indexer.wch <- batch:
+		return nil
+	case <-w.indexer.Dying():
+		return errors.New("writer closed")
+	}
+}
+
+func (w *Writer) Write(rec *zng.Record) error {
+	select {
+	case <-w.indexer.Dying():
+		return errors.New("writer closed")
+	default:
+		return w.WriteCloser.Write(rec)
+	}
+}
+
+func (w *Writer) Close() error {
+	err := w.WriteCloser.Close()
+	w.indexer.Kill(nil)
+	if ierr := w.indexer.Wait(); err == nil && !errors.Is(ierr, context.Canceled) {
+		err = ierr
+	}
+	return err
+}
+
+func (w *Writer) Abort() {
+	w.Close()
+	w.indexer.microindex.Abort()
+}
+
+type indexer struct {
+	tomb.Tomb
```

Comment:
tomb is a package that makes it easier to manage the life span of a goroutine. It is similar to `errgroup` except it includes a dying chan, which signals to the managed goroutine that the parent wants it to exit. This was important in this use case because the Writer needs a way to tell the `indexer` that it has reached the end of the stream.